testng-core/src/main/java/org/testng/internal/invokers/TestInvoker.java

```diff
@@ -232,7 +233,14 @@ public FailureContext retryFailed(
   }
 
   public void runTestResultListener(ITestResult tr) {
-    TestListenerHelper.runTestListeners(tr, m_notifier.getTestListeners());
+    if (tr.getStatus() == ITestResult.STARTED) {
```

Comment:
We should extract out the sorting statement into the `if` condition and the rest of it can reside outside.

Reply:
Looks like this, right?
```
public void runTestResultListener(ITestResult tr) {
	List<ITestListener> testListener = m_notifier.getTestListeners();
    if (tr.getStatus() == ITestResult.STARTED) {
      TestListenerHelper.runTestListeners(tr, testListener);
    } else {
      List<ITestListener> testListenersReversed = testListener;
      Collections.reverse(testListenersReversed);
      TestListenerHelper.runTestListeners(tr, testListenersReversed);
    }
  }
```

Reply:
Not if we do something like this

```java
  public void runTestResultListener(ITestResult tr) {
    boolean isStarted = tr.getStatus() == ITestResult.STARTED;
    List<ITestListener> listeners = isStarted ? m_notifier.getTestListeners()
    : Lists.newReversedArrayList(m_notifier.getTestListeners());
    TestListenerHelper.runTestListeners(tr, listeners);
  }
```

Here again, we would need to add the new method like below into `Lists` as how I mentioned before

```java
  public static <K> List<K> newReversedArrayList(Collection<K> c) {
    List<K> list = newArrayList(c);
    Collections.reverse(list);
    return list;
  }
```

We would need to employ the same technique across so that we can do away with code duplication.